### PR TITLE
AFRL CSV to JSON (new API compatible)

### PR DESCRIPTION
**Pull request by bearmit**
**this PR is a draft until we have the authorization to publish the data, but code can be reviewed from now**

This PR contains the source code (typescript) to process AFRL dataset (CSV) and to produce a JSON compatible with the API.
Related to the task: https://trello.com/c/4eTnhJrp/29-transform-the-afrl-dataset

**Details:**

I explained how this works in a dedicated [README.md](https://github.com/C-Accel-CRIPT/criptscripts/tree/feat/afrl-to-new-api/scripts/typescript/afrl-to-new-api).

The goal of this first pass is to produce a JSON compatible (uploaded [here](https://drive.google.com/drive/folders/1YHXwVf3qscRCqt-pqqqUtXz_jQG22onC) ), but does not guaranty the data is translated 100% accurately.

Here are the problems I was facing, I found work around but I clearly not found "the solution":

- one_phase_direction cannot be stored in CRIPT (+one_phase_direction vocab is not allowed)
- the original script was searching materials from their "cas", I cannot do this, so it may create duplicates between the JSON and the existing database. **solution: we will use chemical_id search once implemented**
- The field Property.value does not accept a number, I had to convert it to string.
- To avoid local duplicated nodes, I used the technique Fatjon show me (using `uid`). It works but the JSON contains 2x more data. There is a possibility to reduce this by removing extra fields for the materials in the inventories, but I am not sure the backend will appreciate.

Feel free to add any comment / suggestions. Thanks.

### CODE REVIEW

**Comment by nh916** [comment]
> First commit just to setup the typescript subproject to turn AFRL data into a JSON compatible with the new API. I just need a feed back regarding to the folder structure I proposed and where I want to add it. I won't merge until the feature is done, but I need your feedback about it.

I don't have any specific organization picked out for this repository when I was working on it, I just went with what I thought was intuitive and clean. I think the organization that you picked out could work, but I was thinking that if the script was only a single page then it might be easier to link to, such as the **bibliography page**
* [CRIPT Script page](https://criptscripts.org/scripts/bibliography%20creator.html)
* [Repository Page](https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/bibliography_creator)

If it needs to be longer, then that is no problem at all either. I would rather it be clean and longer than it be hard to read and shorter.

I think If we decide to put the current organization on CRIPT Script, then we would have to link to the GitHub repository directory where they can find it easily, like the **Simulation Procedure Script**
* [CRIPT Script page](https://criptscripts.org/scripts/simulation_procedure.html)
* [Repository Page](https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/simulation_procedure)

What are your thoughts?

**Comment by bearmit** [comment]
> If it needs to be longer, then that is no problem at all either. I would rather it be clean and longer than it be hard to read and shorter.
> 
> I think If we decide to put the current organization on CRIPT Script, then we would have to link to the GitHub repository directory where they can find it easily, like the **Simulation Procedure Script**
> 
> * [CRIPT Script page](https://criptscripts.org/scripts/simulation_procedure.html)
> * [Repository Page](https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/simulation_procedure)
> 
> What are your thoughts?

Thanks for your feedback.

I am doing closer to your second option, with a link to the repository's sub-folder. The reason is: I am not relying on Python or PythonSDK, but I am instead using TypeScript and reusing the types we defined within the editor to query the new backend, so I have many files.
I could turn the script into a command line interface script easy to run, but the user will need the full folder to install and to run it, or I should publish a package which is really using a bazooka to shot a mosquito. Right now, I am not really coding having in mind people outside of CRIPT will reuse it. I keep existing comment, refactor unclear parts though. But I could do this additional effort later, right now the objective I have is to import the data asap. I might look at the Cript Scripts website after that if I get so time assigned for it.

Thanks again for your help.

**Comment by bearmit** [comment]
First version producing a JSON not rejected by the backend. Curl return success, but the backend does not seems to insert the data.

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 1]
```diff
@@ -0,0 +1,631 @@
+import { AFRLData } from "../types/afrl"
+import { ICitation, ICollection, ICondition, IInventory, IMaterial, IProject, IProperty, IReference } from "../types/cript"
+import csvtojson from "csvtojson";
+
+/**
+ * This script is a typescript port of @see https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/AFRL
+ * The main difference is the data is already well structured to be used in TS (see ../data/data.ts).
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts]: This line is not correct. Data is now read from the CSV.

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+            ] as IInventory[]
+        } as ICollection;
+
+        // Create a project with the collection in it        
+        this.project = {
+            name: config.project_name,
+            model_version: MODEL_VERSION,
+            node: ['Project'],
+            collection: [this.collection],
+        } as IProject;
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:91]: Should add a "material" field with an empty array.

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+        const existing_solvent = this.solvents.get(cas);
+        if (existing_solvent) {
+            console.log(`-- Found existing solvent: ${existing_solvent.name} (cas: ${cas})`)
+            return existing_solvent
+        }
+
+        // Pull solvent from server
+        // Blocker: the script assumes the solvent already exists on the backend, but we probably don't have it.
+        /*
+        try {
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:165]: Here, I skip the search because the API does not allow me to do it yet.

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+            // identifiers.append(cript.Identifier(key="preferred_name", value=name))
+            // Waiting for Brilant's answer I am using names instead.
+            mixture.names.push(name)
+        }
+
+        // Create properties
+        if (conc_vol_fraction && !isNaN(conc_vol_fraction)) {
+            mixture.property.push({
+                key: "conc_vol_fraction",
+                value: String(conc_vol_fraction), // FIXME: backend does not accept numbers,
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:316]: number => string is incorrect, but backend requires it.

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+        }
+
+        if (conc_mass_fraction && !isNaN(conc_mass_fraction)) {
+            mixture.property.push({
+                key: "conc_mass_fraction",
+                value: String(conc_mass_fraction), // FIXME: backend does not accept numbers
+                // "components_relative" does not exist on new API, using "component" instead.         
+                component: [polymer],
+                citation,
+                node: ['Property'],
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:333]: number => string is incorrect, but backend requires it.

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+        // Create identifiers
+        //
+        // The concept of Identifier separate from Material does not exist anymore on the new API.
+        // Instead, a Material has each possible identifier as its own object property.
+        // 
+        if (name) {
+            // There is no such field in the new API
+            // identifiers.append(cript.Identifier(key="preferred_name", value=name))
+            // Waiting for Brilant's answer I am using names instead.
+            mixture.names.push(name)
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:309]: Not certain this value should be assigned to "names".

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+        if (temp_cloud && !isNaN(temp_cloud)) {
+
+            const temp_cloud_property: IProperty = {
+                key: "temp_cloud",
+                value: String(temp_cloud), // FIXME: backend does not accept numbers
+                // "components_relative" does not exist on new API, using "component" instead.         
+                component: [polymer],
+                citation,
+                node: ['Property'],
+                type: 'value',  // FIXME: is this correct from a chemist point of view?
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:348]: Is this correct?

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+                model_version: MODEL_VERSION,
+            } as IProperty)
+        }
+
+        if (temp_cloud && !isNaN(temp_cloud)) {
+
+            const temp_cloud_property: IProperty = {
+                key: "temp_cloud",
+                value: String(temp_cloud), // FIXME: backend does not accept numbers
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:343]: number => string is incorrect, but backend requires it.

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+                condition: [] // will be filled below...
+            } as Partial<IProperty> as any;
+
+            // If present, add conditions
+
+            if (pressure)
+                temp_cloud_property.condition.push({
+                    node: ['Condition'],
+                    key: "pressure",
+                    value: String(pressure), // FIXME: typings are wrong, we should be able to use a number
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:360]: number => string is incorrect, but backend requires it.

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+                    unit: "MPa",
+                    model_version: MODEL_VERSION,
+                } as ICondition);
+
+
+            if (one_phase_direction) {
+
+                // FIXME: uncomment once backend accepts custom vocab (starts with a "+")
+                /*
+                temp_cloud_property.condition.push({
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:370]: How do we deal with this condition key?

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 11]
```diff
@@ -0,0 +1,631 @@
...
+        // hack
+        // In order to avoid to push multiple times the same Node, we have to
+        // set a "uid" (not "uuid"), which is like a local id.
+        const assign_uid = (material: IMaterial) => material.uid = `afrl-csv-to-json:${this.generate_uid()}`;
+        this.solvents.forEach(assign_uid)
+        this.mixtures.forEach(assign_uid)
+        this.polymers.forEach(assign_uid)
+
+        console.log('Loading data OK')
+        return this.project;
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:639]: Major issue, no materials are added to the Project. They are only added to their respective Inventory.

File: scripts/typescript/afrl-to-new-api/README.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+```
+curl -X POST -H "Content-Type: application/json" -d "@./out/afrl-transformed.min.json" <host>/project/ --header "authorization: Bearer <token>"
+```
+
+# How it works?
+
+Step by step of what the script does on a macro level (for more details look at the source code).
+
+- Load the data (`./src/data/*.csv`) as an `Array<AFRLData>`.
+- For each data, we create 3 `Material`s (mixture, solvent, polymer) with some `Property`, `Condition` and `Citation` (`doi` is stored in a `Reference`) on it.
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/README.md]:
```suggestion
- For each CSV line (or object), we create 3 `Material`s (a mixture, a solvent, and a polymer) with some `Property`, `Condition` and `Citation` (`doi` is stored in a `Reference`) on it.
```

File: scripts/typescript/afrl-to-new-api/README.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+- Each type of material is registered in a dedicated `Inventory`.
+- Inventories are store in a single `Collection` which is in a single `Project`.
+- The `Project` is serialized as a single JSON (in two version, one minified and an other human-readable).
+- If some data cannot be transformed, logs are added to the *.errors.json file.
+
+# History
+
+This script was using Python originaly, last commit was from December 2022. I figured out the python version was existing **after** starting this task, that's why I used typescript.
+
+However, I couldn't us Python (no experience) and the PythonSDK (in development), so I decided to continue this way.
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/README.md]:
```suggestion
However, I couldn't use Python (no experience and culture) and the PythonSDK (still in development), so I decided to continue this way.
```

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts (hunk 1 above)
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts]:
```suggestion
 * The main difference is the source data is now read from the original CSV and not a preprocessed *.js.
```

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 14]
```diff
@@ -0,0 +1,631 @@
+import { AFRLData } from "../types/afrl"
+import { ICitation, ICollection, ICondition, IInventory, IMaterial, IProject, IProperty, IReference } from "../types/cript"
+import csvtojson from "csvtojson";
+
+/**
+ * This script is a typescript port of @see https://github.com/C-Accel-CRIPT/criptscripts/tree/master/scripts/python_sdk_scripts/AFRL
+ * The main difference is the data is already well structured to be used in TS (see ../data/data.ts).
+ * The reason why data is a a AFRLData[] is because the legacy frontend (django app) had a JS file with the same data in it.
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts]:
```suggestion
```

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 15]
```diff
@@ -0,0 +1,631 @@
...
+ * Having a *.ts file is very usefull to have autocompletion and static typecheck. *
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts]:
```suggestion
```

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 16]
```diff
@@ -0,0 +1,631 @@
...
+                } as ICondition);
+
+
+            if (one_phase_direction) {
+
+                // FIXME: uncomment once backend accepts custom vocab (starts with a "+")
+                /*
+                temp_cloud_property.condition.push({
+                    node: ['Condition'],
+                    key: "+one_phase_direction", // Not sure this will work, needs custom vocabulary (starts with a "+").
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:372]: How do we deal with this condition key?

File: scripts/typescript/afrl-to-new-api/README.md [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+- Load the data (`./src/data/*.csv`) as an `Array<AFRLData>`.
+- For each CSV line (or object), we create 3 `Material`s (a mixture, a solvent, and a polymer) with some `Property`, `Condition` and `Citation` (`doi` is stored in a `Reference`) on it.
+- Each type of material is registered in a dedicated `Inventory`.
+- Inventories are store in a single `Collection` which is in a single `Project`.
+- The `Project` is serialized as a single JSON (in two version, one minified and an other human-readable).
+- If some data cannot be transformed, logs are added to the *.errors.json file.
+
+# History
+
+This script was using Python originaly, last commit was from December 2022. I figured out the python version was existing **after** starting this task, that's why I used typescript.
```
Comment (nh916) [inline scripts/typescript/afrl-to-new-api/README.md]: I think it might be a good idea to link to the original Python AFRL script for easy find. There is also a typo in `"originaly"`

such as: 

> The script to upload this data to CRIPT had originally been a [python script](https://github.com/C-Accel-CRIPT/cript-ingestion-scripts/tree/master/afrl). The last commit was from December 2022.
> I figured out the python version existed **after** starting this task, which is why I wrote this using typescript.

**Comment by nh916** [comment]
@bearmit I am not sure if we can put this into [CRIPTScripts](https://criptscripts.org/) yet because I don't think we have the full legal clearance from Air Force Research Labs. I think @Ardi028 knows much more on this topic than I do

File: scripts/typescript/afrl-to-new-api/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+const output_dir_path = path.resolve(package_directory_path, './out/');
+const file_name = 'AFRL_linear_polymer_3pdb_data_csv_4_5_2023.csv'; // TODO: read this from command line argument
+
+// Call main as a promise because low level await are not allowed.
+main().then( code => process.exit(code) )
+
+async function main(): Promise<number> {
+
+    console.log('=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-==-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=')
+    console.log('=-=-=-=-=-=-=-=-=-=-=-=-= AFRL CSV to JSON -=-=-=-=-=-==-=-=-=-=-=-=-=')
```
Comment (nh916) [inline scripts/typescript/afrl-to-new-api/src/index.ts:15]: Love this!

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts (hunk 3 above)
Comment (nh916) [reply scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:165]: You may want to talk to @brili about this because I think he may want you to use an exact search from the API and replace all the material UUID with real UUID.

File: scripts/typescript/afrl-to-new-api/package.json [hunk 19]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "afrl-transformation",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "start": "npx ts-node src/index.ts",
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
```
Comment (nh916) [inline scripts/typescript/afrl-to-new-api/package.json]: do you think it would be a good idea to add your name or CRIPT to the authors list?

File: scripts/typescript/afrl-to-new-api/package.json [hunk 20]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "afrl-transformation",
+  "version": "1.0.0",
+  "description": "",
```
Comment (nh916) [inline scripts/typescript/afrl-to-new-api/package.json]: do you think it would be a good idea to add a short description here?

**Comment by nh916** [comment]
Love all the comments and clean code!

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts (hunk 3 above)
Comment (bearmit) [reply scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:165]: The commented code is searching material from a "cas" from CRIPT, I'll do once this is implemented yes.

**Comment by bearmit** [comment]
> @bearmit I am not sure if we can put this into [CRIPTScripts](https://criptscripts.org/) yet because I don't think we have the full legal clearance from Air Force Research Labs. I think @Ardi028 knows much more on this topic than I do

Wait, is this repository public ??

**Comment by bearmit** [comment]
I reopen because the source AFRL data (CSV file) is public. The transformed data won't be added (to the repository and to the website) until we have the authorization.

File: scripts/typescript/afrl-to-new-api/package.json (hunk 19 above)
Comment (bearmit) [reply scripts/typescript/afrl-to-new-api/package.json]: I added my name + CRIPT + MIT. Also set the license to MIT too.

File: scripts/typescript/afrl-to-new-api/README.md (hunk 17 above)
Comment (bearmit) [reply scripts/typescript/afrl-to-new-api/README.md]: > Love all the comments and clean code!

I actually kept a lot of the existing comments coming from the existing version ;)

File: scripts/typescript/afrl-to-new-api/README.md [hunk 21]
```diff
@@ -0,0 +1,36 @@
+# AFRL transformation script
+
+The purpose of this node script is to convert AFRL dataset to a CRIPT compatible JSON.
+Related to: https://trello.com/c/4eTnhJrp/29-transform-the-afrl-dataset
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/README.md]: This should not be merged :)
```suggestion
```

File: scripts/typescript/afrl-to-new-api/package.json [hunk 22]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "afrl-transformation",
+  "version": "1.0.0",
+  "description": "Script to convert an AFRL CSV dataset to a CRIPT compatible JSON",
+  "main": "index.js",
+  "scripts": {
+    "start": "npx ts-node src/index.ts",
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/package.json:8]: In order to produce qualitative data, we should use a testing framework to ensure some edge cases are well handled. I am thinking about:
- How an empty cell (string) is converted to number|string|other? (we want to control that)
- Can we find the location of the errors reported in *..errors.json inside the source CSV? (we store the csv row index, but I think we do not log it in errors messages)
- We should use a JSON validator (ajv) to validate the Project JSON before to send it to the backend
- We could automatize the POST within the typescript script.
- Is smiles_to_bigsmiles working correctly?

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts [hunk 23]
```diff
@@ -0,0 +1,642 @@
...
+        const unique_name = `${name}_${polymer_id}`;
+        const cas = row.polymer_CAS;
+        const bigsmiles = this.smiles_to_BigSMILES(row.polymer_SMILES)
+        const mw_w = row.polymer_Mw;
+        const mw_d = row.polymer_PDI;
+
+        // Try to get the existing
+        if (citation) {
+            // Note: previous implementation was using both mw_w, mw_d, and name as hash, but using unique_name seems more appropriate.
+            const existing_polymer = this.polymers.get(unique_name);
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:214]: I am unsure of that now. Should I use the `cas` as keys in `this.polymer` map like I do for `this.solvents` ?
I case a change is required here, update line 281 accordingly.

File: scripts/typescript/afrl-to-new-api/README.md [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+```
+
+# How it works?
+
+Step by step of what the script does on a macro level (for more details look at the source code).
+
+- Load the data (`./src/data/*.csv`) as an `Array<AFRLData>`.
+- For each CSV line (or object), we create 3 `Material`s (a mixture, a solvent, and a polymer) with some `Property`, `Condition` and `Citation` (`doi` is stored in a `Reference`) on it.
+- Each type of material is registered in a dedicated `Inventory`.
+- Inventories are store in a single `Collection` which is in a single `Project`.
```
Comment (bearmit) [inline scripts/typescript/afrl-to-new-api/README.md]:
```suggestion
- Inventories are stored in a unique `Collection` attached to a unique `Project`.
```

**Comment by nh916** [comment]
> > @bearmit I am not sure if we can put this into [CRIPTScripts](https://criptscripts.org/) yet because I don't think we have the full legal clearance from Air Force Research Labs. I think @Ardi028 knows much more on this topic than I do
> 
> Wait, is this repository public ??

yes it is

File: scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts (hunk 23 above)
Comment (nh916) [reply scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:214]: From what I remember `CAS` was renamed to `chemical_id` for our platform

Comment (bearmit) [reply scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:214]: There is `cas` on `Material` schema. It is described in the `material_identifier_key ` controlled vocab too.

Comment (nh916) [reply scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:214]: it should have been replaced with `chemical_id` because `CAS` is unavailable to our platform

Comment (bearmit) [reply scripts/typescript/afrl-to-new-api/src/scripts/afrl-csv-to-json.ts:214]: You are correct. The vocabulary do not mention `cas` but do mention `chemical_id`. I'll still use `cas` in the code because the source data uses this concept.
Unfortunately I cannot store the `cas` as a `chemical_id` on a material with the current schema. Such a field does not exists. Since `cas` is present on the schema I fallback on this field temporarly:

image

**Comment by bearmit** [comment]
I am right now working on an other transformation (RCBC), I am based on this branch and had to refactor. I introduced a `CriptJSON.stringify` class (acting like `JSON.stringify`), to optimize the JSON file size (file size can be divided up to 4x). File could also be send compressed as binary.
See PR #6

**Comment by bearmit** [comment]
I closed this PR, I reorganized my branches to have a clean history and a common branch for another script (in typescript too).
new PR: AFRL #8 
other script: RCBC #9
